HandEyeCalibration/Transform/Transform.cpp

```diff
@@ -0,0 +1,330 @@
+                break;
+            case ZYX_INTRINSIC:
+            case XYZ_EXTRINSIC:
+                rotationMatrix = Eigen::AngleAxisd(rollPitchYaw[2], Eigen::Vector3d::UnitZ())
+                                 * Eigen::AngleAxisd(rollPitchYaw[1], Eigen::Vector3d::UnitY())
+                                 * Eigen::AngleAxisd(rollPitchYaw[0], Eigen::Vector3d::UnitX());
+                break;
+            default: throw "Invalid orientation";
+        }
+
+        return rotationMatrix;
+    }
+
+    int compareRotation(const Eigen::Matrix3d rotationMatrix1, const Eigen::Matrix3d rotationMatrix2)
+    {
+        std::cout << "\nAbsolute error introduced by transformation" << std::endl;
+        const auto rotationMatrixError = (rotationMatrix1 - rotationMatrix2).cwiseAbs();
+        std::cout << rotationMatrixError << std::endl;
+
+        std::cout << "\nRelative (elementwise percent) error introduced by transformation" << std::endl;
+        std::cout << (rotationMatrixError.cwiseQuotient(rotationMatrix1)) * 100 << std::endl;
+
+        std::cout << "\nRotation difference introduced by transformation" << std::endl;
+        const auto differenceRotationMatrix = rotationMatrix1 * rotationMatrix2.inverse();
+        std::cout << differenceRotationMatrix << std::endl;
+
+        std::cout << "\nRotation distance introduced by transformation" << std::endl;
+        const auto axisAngle = Eigen::AngleAxisd(differenceRotationMatrix);
+        auto rotationDistance = axisAngle.angle();
+        std::cout << rotationDistance << std::endl;
+
+        const auto quaternion1 = Eigen::Quaterniond(rotationMatrix1);
+        const auto quaternion2 = Eigen::Quaterniond(rotationMatrix2);
+        rotationDistance = 2 * acos((quaternion1 * quaternion2.conjugate()).w());
+        std::cout << rotationDistance << std::endl;
+
+        return 1;
+    }
+
+    int compareRotationExtended(const Eigen::Array3d rollPitchYaw,
+                                const Eigen::Matrix3d rotationMatrix,
+                                const TransformUtils::Rotation rotation)
+    {
+        const auto rollPitchYawNew = TransformUtils::rotmat2rpy(rotationMatrix, rotation);
+
+        std::cout << "\nOriginal RollPitchYaw:\n" << rollPitchYaw << std::endl;
+        std::cout << "\nNew RollPitchYaw:\n" << rollPitchYawNew << std::endl;
+
+        const auto rotationMatrixNew = TransformUtils::rpy2rotmat(rollPitchYawNew, rotation);
+
+        std::cout << "Original Rotation Matrix:\n" << rotationMatrix << std::endl;
+        std::cout << "New Rotation Matrix:\n" << rotationMatrixNew << std::endl;
+
+        // Calculate error in transformation (Rotation Matrix --> Roll-Pitch-Yaw --> Rotation Matrix)
+        std::cout << "\nVerifying Rotation Matrix" << std::endl;
+        compareRotation(rotationMatrix, rotationMatrixNew);
+
+        return 1;
+    }
+
+} // namespace TransformUtils
+
+int main()
```

Comment:
The original idea was to create convenience functions to use in other samples. 
Then we found that we wanted main() to show how these functions should be used... Is there a way in C++ (like in Python) to have "main()" present in a module, that is only compiled if the module is the "main file"?
I'm thinking about something like:
```
if __name__ == "__main__":
    _main()
```
I guess the answer is that we should separate into two files? What we discussed (in CSU) was to keep it like this for now. Customers would probably only require one or two of the conversions and could copy/paste the code. Maybe later it would make sense to have a Class with these convenience functions... 
For this reason I guess we could put main() on top in this file as well, to be consistent with other sample applications.

Reply:
If you want to factor out some convenience functions, I would put them in a separate file without `main`. Then if you want an executable to demo them, do that as a separate file.

Normally we're very strict about not duplicating functionality and instead extracting helpers, but for a sample repository there are other considerations. There are some advantages to having everything available in one file, not having to look up other files to find functions etc. So for this repo, I wouldn't worry too much about a bit of duplication.